client/components/isomorphic-id/index.js

```diff
@@ -0,0 +1,42 @@
+/** @ssr-ready **/
+
+/**
+ * External dependencies
+ */
+import React from 'react';
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
+
+/**
+ * Internal dependencies
+ */
+import { incrementIsomorphicId } from 'state/ui/actions';
+import { getCurrentIsomorphicId } from 'state/ui/selectors';
+
+export default function addUniqueComponentId( Component ) {
+	class ComponentWithId extends React.Component {
+		constructor( props ) {
+			super( props );
+
+			this.componentId = this.props.currentIsomorphicId;
+			this.props.incrementIsomorphicId();
```

Comment:
I think this might be ok due to [run to completion](https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop#.22Run-to-completion.22) but be wary in any other language for reader/writer problems